File: app/src/main/java/com/pr0gramm/app/services/UpdateChecker.java
package com.pr0gramm.app.services;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.FileProvider;

import com.google.common.collect.ImmutableList;
import com.google.common.io.ByteStreams;
import com.google.gson.GsonBuilder;
import com.pr0gramm.app.AppComponent;
import com.pr0gramm.app.BuildConfig;
import com.pr0gramm.app.Dagger;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.ui.fragments.DownloadUpdateDialog;
import com.pr0gramm.app.util.AndroidUtility;
import com.pr0gramm.app.util.BackgroundScheduler;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Actions;
import rx.util.async.Async;

import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;

/**
 * Class to perform an update check.
 */
public class UpdateChecker {
    private static final Logger logger = LoggerFactory.getLogger("UpdateChecker");

    private final int currentVersion;
    private final ImmutableList<String> endpoints;

    public UpdateChecker(Context context) {
        currentVersion = AndroidUtility.buildVersionCode();

        boolean betaChannel = Settings.get().getUseBetaChannel();
        endpoints = updateUrls(betaChannel);
    }

    private Observable<Update> check(String endpoint) {
        return Async.fromCallable(() -> {
            UpdateApi api = newRestAdapter(endpoint).create(UpdateApi.class);
            return api.get().execute().body();

        }, BackgroundScheduler.instance()).filter(update -> {
            logger.info("Installed v{}, found update v{} at {}",
                    currentVersion, update.version(), endpoint);

            // filter out if up to date
            return update.version() > currentVersion;
        }).map(update -> {
            // rewrite url to make it absolute
            String apk = update.apk();
            if (!apk.startsWith("http")) {
                apk = Uri.withAppendedPath(Uri.parse(endpoint), apk).toString();
            }

            logger.info("Got new update at url " + apk);
            return ImmutableUpdate.builder()
                    .version(update.version())
                    .changelog(update.changelog())
                    .apk(apk)
                    .build();
        });
    }

    public Observable<Update> check() {
        return Observable.from(endpoints)
                .flatMap(ep -> check(ep)
                        .doOnError(err -> logger.warn("Could not check for update at {}: {}", ep, err.toString()))
                        .onErrorResumeNext(Observable.empty()))
                .take(1);
    }

    private static Retrofit newRestAdapter(String endpoint) {
        com.google.gson.Gson gson = new GsonBuilder()
                .registerTypeAdapterFactory(new GsonAdaptersUpdate())
                .create();

        return new Retrofit.Builder()
                .baseUrl(endpoint)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
    }

    private interface UpdateApi {
        @GET("update.json")
        Call<Update> get();
    }

    /**
     * Returns the Endpoint-URL that is to be queried
     */
    private static ImmutableList<String> updateUrls(boolean betaChannel) {
        List<String> urls = new ArrayList<>();

        if (betaChannel) {
            urls.add("https://pr0.wibbly-wobbly.de/beta/");
            urls.add("https://github.com/mopsalarm/pr0gramm-updates/raw/beta/");
        } else {
            urls.add("https://pr0.wibbly-wobbly.de/stable/");
            urls.add("https://github.com/mopsalarm/pr0gramm-updates/raw/master/");
        }

        return ImmutableList.copyOf(urls);
    }


    public static void download(FragmentActivity activity, Update update) {
        AppComponent appComponent = Dagger.appComponent(activity);
        Observable<DownloadService.Status> progress = appComponent.downloadService()
                .downloadToFile(update.apk())
                .subscribeOn(BackgroundScheduler.instance())
                .unsubscribeOn(BackgroundScheduler.instance())
                .observeOn(AndroidSchedulers.mainThread())
                .share();

        // install on finish
        Context appContext = activity.getApplicationContext();
        progress.filter(DownloadService.Status::getFinished)
                .flatMap(status -> {
                    try {
                        install(appContext, status.getFile());
                        return Observable.empty();

                    } catch (IOException error) {
                        return Observable.error(error);
                    }
                })
                .subscribe(Actions.empty(), defaultOnError());

        // show a progress dialog
        DownloadUpdateDialog dialog = new DownloadUpdateDialog(progress);
        dialog.show(activity.getSupportFragmentManager(), null);

        // remove pending upload notification
        appComponent.notificationService().cancelForUpdate();
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private static void install(Context context, File apk) throws IOException {
        Uri uri;
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            String provider = BuildConfig.APPLICATION_ID + ".FileProvider";
            uri = FileProvider.getUriForFile(context, provider, apk);
        } else {
            File file = new File(context.getExternalCacheDir(), "update.apk");

            logger.info("Copy apk to public space.");
            try (InputStream input = new FileInputStream(apk)) {
                try (OutputStream output = new FileOutputStream(file)) {
                    ByteStreams.copy(input, output);
                }
            }

            // make file readable
            if (!file.setReadable(true))
                logger.info("Could not make file readable");

            uri = Uri.fromFile(file);
        }

        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setDataAndType(uri, "application/vnd.android.package-archive");
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        context.startActivity(intent);
    }
}

File: app/src/main/java/com/pr0gramm/app/services/UpdateChecker.kt
package com.pr0gramm.app.services

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import androidx.core.content.FileProvider
import androidx.core.net.toFile
import androidx.fragment.app.FragmentActivity
import com.pr0gramm.app.BuildConfig
import com.pr0gramm.app.Logger
import com.pr0gramm.app.MoshiInstance
import com.pr0gramm.app.Settings
import com.pr0gramm.app.model.update.UpdateModel
import com.pr0gramm.app.ui.Screen
import com.pr0gramm.app.ui.base.BaseAppCompatActivity
import com.pr0gramm.app.ui.base.launchUntilDestroy
import com.pr0gramm.app.ui.fragments.ProgressDialogController
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.di.injector
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.onEach
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.create
import retrofit2.http.GET
import retrofit2.http.Query
import retrofit2.http.Url
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

/**
 * Class to perform an update check.
 */
class UpdateChecker {
    private val currentVersion = AndroidUtility.buildVersionCode()

    private val updateApi = Retrofit.Builder()
            .baseUrl("https://example.com")
            .addConverterFactory(MoshiConverterFactory.create(MoshiInstance))
            .validateEagerly(BuildConfig.DEBUG)
            .build().create<UpdateApi>()

    private val endpoints: List<String> = mutableListOf<String>().also { urls ->
        if (Settings.useBetaChannel) {
            urls += "https://app.pr0gramm.com/updates/beta/update.json"
            urls += "https://raw.githubusercontent.com/pr0gramm-com/pr0gramm-app/refs/heads/updates/beta.json"
        } else {
            urls += "https://app.pr0gramm.com/updates/stable/update.json"
            urls += "https://raw.githubusercontent.com/pr0gramm-com/pr0gramm-app/refs/heads/updates/stable.json"
        }
    }

    private suspend fun queryOne(endpoint: String): Update {
        val update = Update(updateApi.fetchUpdateAsync(
                endpoint, androidVersion = Build.VERSION.SDK_INT))

        // make path absolute if needed
        var apk = update.apk
        if (!apk.startsWith("http")) {
            apk = Uri.withAppendedPath(Uri.parse(endpoint), apk).toString()
        }

        // use this one as our update.
        return update.copy(apk = apk)
    }

    suspend fun queryAll(): Response {
        // start all jobs in parallel
        val tasks = supervisorScope {
            endpoints.map { async { queryOne(it) } }
        }

        // wait for them to finish
        val results = tasks.map { runCatching { it.await() } }

        if (results.all { it.isFailure }) {
            val err = results.first().exceptionOrNull() ?: return Response.NoUpdate
            return Response.Error(err)
        }

        val update = results
                .mapNotNull { it.getOrNull() }.maxByOrNull { it.version }
                ?.takeIf { it.version > currentVersion }
                ?: return Response.NoUpdate

        return Response.UpdateAvailable(update)
    }

    sealed class Response {
        object NoUpdate : Response()

        class Error(val err: Throwable) : Response()

        class UpdateAvailable(val update: Update) : Response()
    }

    private interface UpdateApi {
        @GET
        suspend fun fetchUpdateAsync(@Url url: String, @Query("androidVersion") androidVersion: Int): UpdateModel
    }

    companion object {
        private val logger = Logger("UpdateChecker")

        fun download(activity: FragmentActivity, update: Update) {
            // rotating crashes the dialog
            Screen.lockOrientation(activity)

            val downloadService = activity.injector.instance<DownloadService>()
            val notificationService = activity.injector.instance<NotificationService>()

            activity as BaseAppCompatActivity

            activity.launchUntilDestroy {

                val progress = downloadService.downloadUpdateFile(Uri.parse(update.apk))

                // show the progress dialog.
                val dialog = ProgressDialogController(activity)

                dialog.show()

                val fileUri = try {
                    // download the file and show progress as long as the dialog is open
                    progress.onEach { status -> dialog.updateStatus(status) }
                            .firstOrNull { state -> state.file != null }
                            ?.file

                } finally {
                    // remove dialog again
                    dialog.dismiss()
                }

                if (fileUri != null) {
                    install(activity, fileUri.toFile())
                }
            }

            // remove pending upload notification
            notificationService.cancelForUpdate()
        }

        @SuppressLint("SdCardPath")
        private suspend fun install(context: Context, apk: File) {
            val uri = withContext(Dispatchers.IO + NonCancellable) {
                if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
                    val provider = BuildConfig.APPLICATION_ID + ".FileProvider"
                    FileProvider.getUriForFile(context, provider, apk)

                } else {
                    val candidates = listOf(
                            { context.externalCacheDir },
                            { Environment.getExternalStorageDirectory() },
                            { File("/sdcard") }
                    )


                    val directory = candidates
                            .mapNotNull { runCatching(it).getOrNull() }
                            .firstOrNull { dir -> dir.canWrite() }
                            ?: throw IOException("Could not find a public place to write the update to")

                    val file = File(directory, "update.apk")

                    logger.info { "Copy apk to public space." }
                    FileInputStream(apk).use { input ->
                        FileOutputStream(file).use { output ->
                            input.copyTo(output)
                        }
                    }

                    // make file readable
                    if (!file.setReadable(true)) {
                        logger.info { "Could not make file readable" }
                    }

                    Uri.fromFile(file)
                }
            }

            val intent = Intent(Intent.ACTION_VIEW)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            intent.setDataAndType(uri, "application/vnd.android.package-archive")
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            context.startActivity(intent)
        }
    }
}
